src/org/labkey/remoteapi/Command.java

```diff
@@ -129,26 +108,13 @@ public String getActionName()
     }
 
     /**
-     * Returns the current parameter map, or null if a map has not yet been set.
-     * Derived classes will typically override this method to provide a parameter
-     * map based on data members set by specialized setter methods.
-     * @return The current parameter map.
+     * Returns a new, mutable parameter map. Derived classes will typically override this
+     * method to put values passed to specialized setter methods into the map.
+     * @return The parameter map to use when building the URL.
      */
     public Map<String, Object> getParameters()
     {
-        if (null == _parameters)
-            _parameters = new HashMap<>();
-
-        return _parameters;
-    }
-
-    /**
-     * Sets the current parameter map.
-     * @param parameters The parameter map to use
-     */
-    public void setParameters(Map<String, Object> parameters)
-    {
-        _parameters = parameters;
+        return new HashMap<>();
```

Comment:
Should we use an unmodifiable map to fail hard if people try to use the old `getParameters().put(..)` pattern?
```suggestion
        return Collections.emptyMap();
```

`Command.createURI` would need to deal with this and subclasses would need to not build off of `super.getParameters`.

Reply:
`protected createParameterMap()` returns a mutable Map. `public final getParameters()` now returns an immutable copy.